### ISSUE TITLE
Move PRs to `TO REBASE` column in case of conflicts

### ISSUE THREAD

**Issue by churik** [status-im/status-react#13497]
# Problem

PR is moved automatically in https://github.com/status-im/status-react/projects/7 from any column to Contributor  if there are conflicts in PRs (like here https://github.com/status-im/status-react/pull/13358#event-6796108654 )

## Implementation
Change workflow for `status-github-bot` to move PRs with conflicts to the [`TO REBASE` ](https://github.com/status-im/status-react/projects/7#column-18916675) column instead of [`CONTRIBUTOR`](https://github.com/status-im/status-react/projects/7#column-1843046)

**Comment by churik** [comment]
cc @Marudny

**Comment by Marudny** [comment]
Hey, apologies for delays. I was off for a week. I'm during research Github Actions and Github Bot. I aim to close that issue till the end of the week.

**Comment by jakubgs** [comment]
Generally the bot code is here: https://github.com/status-im/status-github-bot

The configuration is here:
https://github.com/status-im/status-mobile/blob/28a2113874cc3bcce370d913cf0c3dd70c85b9ef/.github/github-bot.yml#L3-L9

~~You should be trying to copy one of the existing flows and adjusting it, for example:~~
https://github.com/status-im/status-github-bot/blob/master/bot_scripts/assign-new-pr-to-review.js

**Comment by jakubgs** [comment]
Actually, the flow we want to modify is:
https://github.com/status-im/status-github-bot/blob/master/bot_scripts/assign-approved-pr-to-test.js

Specifically:
```js
      if (process.env.DRY_RUN || process.env.DRY_RUN_PR_TO_TEST) {
        robot.log.info(`${botName} - Would have moved card ${existingGHCard.id} to ${dstColumn.name} for PR #${prInfo.number}`)
      } else {
        // Found in the source column, let's move it to the destination column
        await github.projects.moveProjectCard({id: existingGHCard.id, position: 'bottom', column_id: dstColumn.id})

        robot.log.info(`${botName} - Moved card ${existingGHCard.id} to ${dstColumn.name} for PR #${prInfo.number}`)
      }
```
https://github.com/status-im/status-github-bot/blob/bf35bd6c/bot_scripts/assign-approved-pr-to-test.js#L158-L165

**Comment by jakubgs** [comment]
I see this is the section which decides what state directs to which column:
```js
function getColumns (state, columns) {
  switch (state) {
    case 'awaiting_reviewers':
      return { srcColumns: [columns.contributor, columns.test], dstColumn: columns.review }
    case 'changes_requested':
      return { srcColumns: [columns.review, columns.test], dstColumn: columns.contributor }
    case 'failed':
      return { srcColumns: [columns.review, columns.test], dstColumn: columns.contributor }
    case 'approved':
      return { srcColumns: [columns.contributor, columns.review], dstColumn: columns.test }
    default:
      return { srcColumns: [], dstColumn: null }
  }
}
```
https://github.com/status-im/status-github-bot/blob/bf35bd6c/bot_scripts/assign-approved-pr-to-test.js#L207-L220

Which gets `state` from `_getReviewApprovalState()`:
```js
  const approvedReviews = finalReviews.filter(reviewState => reviewState === 'APPROVED')
  if (approvedReviews.length >= threshold) {
    const reviewsWithChangesRequested = finalReviews.filter(reviewState => reviewState === 'CHANGES_REQUESTED')
    if (reviewsWithChangesRequested.length === 0) {
      robot.log.debug(`No changes requested, considering state as approved`)
      return 'approved'
    }

    robot.log.debug(`${reviewsWithChangesRequested.length} changes requested, considering state as changes_requested`)
    return 'changes_requested'
  }

  robot.log.debug(`Not enough reviewers yet, considering state as awaiting_reviewers`)
  return 'awaiting_reviewers'
```
https://github.com/status-im/status-github-bot/blob/bf35bd6c/lib/github-helpers.js#L84-L97

**Comment by jakubgs** [comment]
I think this is what we need:
```js
  switch (pullRequest.mergeable_state) {
    case 'clean':
      if (testedPullRequestLabelName !== null && pullRequest.labels.find(label => label.name === testedPullRequestLabelName)) {
        robot.log.debug(`Pull request is labeled '${testedPullRequestLabelName}', ignoring`)
        return null
      }

      state = 'approved'
      break
    case 'dirty':
      state = 'failed'
      break
    case 'unstable':
      if (filterIgnoredStatusContextFn) {
        const isSuccess = await _isPullRequestStatusSuccessIgnoringContext(context, filterIgnoredStatusContextFn, pullRequest)
        if (isSuccess) {
          state = 'approved'
          robot.log.debug(`All important statuses are successful, so considering state as ${state}`)
        }
      }
      break
  }
```
https://github.com/status-im/status-github-bot/blob/bf35bd6c/lib/github-helpers.js#L51-L72

Specifically the `mergeable_state` from `pullRequest` object is what we want. But not sure exact meaning of `dirty`.

**Comment by jakubgs** [comment]
We can find `mergeable_state` in these API docs: https://docs.github.com/en/rest/pulls/pulls

But this issue in Octobox.net explains the meaning:

* https://github.com/octokit/octokit.net/issues/1763

> - `dirty` - Merge conflict. Merging will be blocked
> - `unknown` - Mergeability was not checked yet. Merging will be blocked
> - `blocked` - Blocked by a failing/missing required status check.
> - `behind` - Head branch is behind the base branch. Only if required status checks is enabled but loose policy is not. Merging will be blocked.
> - `unstable` - Failing/pending commit status that is not part of the required status checks. Merging is allowed (yellow box).
> - `has_hooks` - GitHub Enterprise only, if a repo has custom pre-receive hooks. Merging is allowed (green box).
> - `clean` - No conflicts, everything good. Merging is allowed (green box).

So yes, `dirty` is what we need.

**Comment by jakubgs** [comment]
Apparently `mergeable_state` is an "undocummented field": https://github.com/orgs/community/discussions/24299

**Comment by jakubgs** [comment]
Fucking finally: https://github.com/status-im/status-mobile/pull/14088

![image](https://user-images.githubusercontent.com/2212681/193264071-4c7062b5-c68c-473f-b27d-b32da2902c0b.png)

**Comment by jakubgs** [comment]
Looks like bot restarts tirgger unnecessary builds sometimes:

* https://github.com/status-im/status-mobile/issues/14089

**Comment by jakubgs** [comment]
Reverted in:

* https://github.com/status-im/status-github-bot/pull/60
